kernel/aster-nix/src/process/process_vm/heap.rs

```diff
@@ -72,18 +63,23 @@ impl Heap {
                     // FIXME: should we allow shrink current user heap?
                     return Ok(current_heap_end);
                 }
+                let old_size = (current_heap_end - self.base).align_up(PAGE_SIZE);
                 let new_size = (new_heap_end - self.base).align_up(PAGE_SIZE);
-                let heap_mapping = root_vmar.get_vm_mapping(USER_HEAP_BASE)?;
-                let heap_vmo = heap_mapping.vmo();
-                heap_vmo.resize(new_size)?;
+
+                // FIXME: Our current implementation of mapping resize cannot move
+                // existing mappings within the new range, which may cause the resize
+                // operation to fail. Therefore, if there are already mappings within
+                // the heap expansion range, the brk operation will fail.
+                root_vmar.resize_mapping(self.base, old_size, new_size)?;
```

Comment:
Why `resize_mapping` will accept a parameter as `old_size`, the vmar should stores the size of existing mappings ,right?

Reply:
Because the `map_addr` parameter may be not a `map_addr` of a `VmMapping`. 
e.g. We have a VmMapping 0x1000-0x3000, and then protect 0x1000-0x2000 to split this mapping into `0x1000-0x2000` and `0x2000-0x3000`. Now we can still invoke `resize_mapping` with parameters 0x1500(map_addr), 0x1500(old_size),  0x2000(new_size), which will resize the second mapping to `0x2000-0x3500`. `mremap` syscall can do this and since we may use `resize_mapping` to implement `mremap`, I think we should keep such parameters.